docs/architecture-decision-records/20220121-conduit-plugin-architecture.md

```diff
@@ -0,0 +1,322 @@
+  except that in this case Conduit is sending records to the plugin and receiving acknowledgments back. Both streams are
+  independent and are able to transmit data concurrently. The plugin is expected to send an acknowledgment back to
+  Conduit for every record it received, even if the record was not processed successfully (in that case the
+  acknowledgment should contain the error). The stream should stay open until either an error occurs or the `Stop`
+  function is called *and* all remaining acknowledgments are sent (this is handled by the SDK).
+- `Stop` - Conduit signals to the plugin that there be no more records written to the request stream in `Run`. The
+  plugin needs to flush any records that are cached in memory, send back the acknowledgments and stop the `Run`
+  function.
+- `Teardown` - see the description of `Teardown` for the source plugin (same purpose).
+
+Last but not least, the specifier interface:
+
+```protobuf
+service SpecifierPlugin {
+  rpc Specify(Specifier.Specify.Request) returns (Specifier.Specify.Response);
+}
+```
+
+This one is very simple and allows the plugin to describe itself. Through this Conduit will get to know the name of the
+plugin, version, author, description and data about what parameters the plugin expects in the configuration.
+
+### Conduit internal plugin interface
+
+Conduit internally defines its own plugin interfaces:
+
+```go
+type SourcePlugin interface {
+  Configure(context.Context, map[string]string) error
+  Start(context.Context, record.Position) error
+  Read(context.Context) (record.Record, error)
+  Ack(context.Context, record.Position) error
+  Stop(context.Context) error
+  Teardown(context.Context) error
+}
+
+type DestinationPlugin interface {
+  Configure(context.Context, map[string]string) error
+  Start(context.Context) error
+  Write(context.Context, record.Record) error
+  Ack(context.Context) (record.Position, error)
+  Stop(context.Context) error
+  Teardown(context.Context) error
+}
+
+type SpecifierPlugin interface {
+  Specify() (Specification, error)
+}
+```
+
+The methods roughly map to the methods defined in the gRPC interface described in the previous chapter. There are 2
+things to point out
+
+- These interfaces use internal structs in the Conduit codebase (e.g. `record.Position`, `record.Record`), so
+  interacting with a plugin is seamless.
+- The bidirectional stream method `Run` is broken down into two separate methods, one for receiving messages from the
+  stream and one to send messages. Those methods can only be called after a call to `Start` since that is the method in
+  which the stream actually gets opened.
+    - In `SourcePlugin` we can read records from the stream by calling `Read`. This method will block until either an
+      error occurs or a new record is produced by the plugin. Successfully processed records can be acknowledged by
```

Comment:
We are free to do so _if we need it_, this is the internal interface in Conduit, the nice thing is that we can change it without impacting plugins at all. The key phrase is "_if we need it_", and right now I think we simply don't need such a function. The source node (which essentially initiates the `Read` call) is now much simpler, it just calls `Read` and waits forever, it will either get a record or an error, making it async would make the logic more complicated.

Reply:
IIUC, the CDC stuff in the Pg plugin is reading events asynchronously (@dylanlott can correct me if I'm wrong). But yeah, if that's the only use case, it may not be worth it.

Reply:
I want to reiterate that this text refers to the internal interface that Conduit uses to communicate with the plugin, it is tailored to the needs we have internally in `stream.SourceNode`, no plugin will ever know about these methods. All the plugin cares about is the interface defined in the SDK, and there it's already designed in a way that the plugin can fetch records however it wants to (i.e. synchronously or asynchronously).